File: backend/app/schemas/generator.py
"""
Generator Schemas

프론트엔드와 공유되는 Generator API 인터페이스

작성일: 2025-11-17
"""

from pydantic import BaseModel, Field
from typing import Any, Dict, Optional, List
from datetime import datetime


class GenerateRequest(BaseModel):
    """
    Generator 실행 요청

    프론트엔드가 POST /api/v1/generate에 전송하는 요청
    """
    kind: str = Field(
        ...,
        description="생성 타입",
        example="product_detail"
    )
    brandId: str = Field(
        ...,
        description="브랜드 ID",
        example="brand_demo"
    )
    input: Dict[str, Any] = Field(
        ...,
        description="입력 데이터 (kind에 따라 다름)"
    )
    options: Optional[Dict[str, Any]] = Field(
        None,
        description="추가 옵션"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "kind": "product_detail",
                "brandId": "brand_demo",
                "input": {
                    "product_name": "딥그린 진정 토너",
                    "features": ["저자극", "지성피부"],
                    "target_audience": "2030 직장인"
                },
                "options": {
                    "tone": "professional",
                    "length": "medium"
                }
            }
        }


class DocumentPayload(BaseModel):
    """Editor Document 구조"""
    documentId: str = Field(..., description="Document ID")
    type: str = Field(..., description="Document 타입")
    canvas_json: Dict[str, Any] = Field(
        ...,
        description="Fabric.js Canvas JSON"
    )


class ImagePayload(BaseModel):
    """생성된 이미지 데이터"""
    type: str = Field(..., description="이미지 타입: 'base64' 또는 'url'")
    format: str = Field(default="png", description="이미지 포맷 (png, jpg 등)")
    data: Optional[str] = Field(None, description="Base64 인코딩된 이미지 데이터 (type='base64'일 때)")
    url: Optional[str] = Field(None, description="이미지 URL (type='url'일 때)")

    class Config:
        json_schema_extra = {
            "example": {
                "type": "base64",
                "format": "png",
                "data": "iVBORw0KGgoAAAANSUhEUgAA..."
            }
        }


class TextPayload(BaseModel):
    """생성된 텍스트 블록"""
    headline: Optional[str] = Field(None, description="헤드라인")
    subheadline: Optional[str] = Field(None, description="서브헤드라인")
    body: Optional[str] = Field(None, description="본문")
    bullets: Optional[List[str]] = Field(None, description="불릿 포인트")
    cta: Optional[str] = Field(None, description="Call to Action")
    image: Optional[ImagePayload] = Field(None, description="생성된 이미지 (선택적)")


class GenerateResponse(BaseModel):
    """
    Generator 실행 응답

    프론트엔드가 받는 응답
    """
    kind: str = Field(..., description="생성 타입")
    document: DocumentPayload = Field(..., description="Editor Document")
    text: TextPayload = Field(..., description="생성된 텍스트")
    meta: Dict[str, Any] = Field(
        default_factory=dict,
        description="메타데이터 (워크플로우 추적, 토큰 사용량 등)"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "kind": "product_detail",
                "document": {
                    "documentId": "doc_xyz789",
                    "type": "product_detail",
                    "canvas_json": {
                        "version": "5.3.0",
                        "objects": []
                    }
                },
                "text": {
                    "headline": "완벽한 소음 차단의 시작",
                    "body": "프리미엄 노이즈캔슬링 기술...",
                    "bullets": ["24시간 배터리", "IPX7 방수"],
                    "image": {
                        "type": "base64",
                        "format": "png",
                        "data": "iVBORw0KGgoAAAANSUhEUgAA..."
                    }
                },
                "meta": {
                    "workflow": "product_content_pipeline",
                    "agents_used": ["copywriter", "reviewer"],
                    "elapsed_seconds": 12.35,
                    "tokens_used": 350
                }
            }
        }

File: frontend/types/generator.ts
/**
 * Generator API Types
 *
 * 백엔드 이미지 생성 API 응답 타입 정의
 * - Base64 및 URL 형식 모두 지원
 * - 나중에 Base64 → URL 전환 시에도 타입 변경 최소화
 *
 * @author C팀 (Frontend Team)
 * @version 1.0
 * @date 2025-11-22
 */

// ============================================================================
// 이미지 관련 타입
// ============================================================================

/**
 * 생성된 이미지 데이터
 * - type: "base64" | "url"
 * - Base64: 초기 구현 (ComfyUI 직접 반환)
 * - URL: 향후 구현 (S3/CDN 업로드 후 URL 반환)
 */
export type GeneratedImage = {
  /** 이미지 데이터 타입 */
  type: 'base64' | 'url';

  /** 이미지 포맷 (png, jpg, jpeg 등) */
  format?: 'png' | 'jpg' | 'jpeg' | string;

  /** Base64 인코딩된 이미지 데이터 (type === "base64"일 때) */
  data?: string;

  /** 이미지 URL (type === "url"일 때) */
  url?: string;
};

// ============================================================================
// 생성 요청 타입
// ============================================================================

/**
 * 제품 콘텐츠 생성 요청
 */
export type ProductContentRequest = {
  /** Agent 역할 */
  role: 'copywriter' | 'designer' | 'reviewer';

  /** 작업 타입 */
  task: 'product_detail' | 'ad_campaign' | 'social_media';

  /** 요청 페이로드 */
  payload: {
    /** 제품 이름 */
    product_name: string;

    /** 제품 특징 */
    features?: string[];

    /** 타겟 고객 */
    target_audience?: string;

    /** 톤앤매너 */
    tone?: string;

    /** 이미지 생성 포함 여부 (기본값: false) */
    include_image?: boolean;

    /** 브랜드 ID */
    brand_id?: string;
  };
};

// ============================================================================
// 생성 응답 타입
// ============================================================================

/**
 * 제품 콘텐츠 생성 응답
 */
export type ProductContentResponse = {
  /** 헤드라인 */
  headline: string;

  /** 서브헤드라인 (선택적) */
  subheadline?: string;

  /** 본문 */
  body: string;

  /** 불릿 포인트 (선택적) */
  bullets?: string[];

  /** Call to Action */
  cta: string;

  /** 생성된 이미지 (include_image=true일 때만) */
  image?: GeneratedImage | null;

  /** 사용량 정보 */
  usage?: {
    tokens?: number;
    cost?: number;
  };

  /** 메타데이터 */
  meta?: {
    agent?: string;
    task?: string;
    timestamp?: string;
  };
};

/**
 * SNS 포스트 생성 응답
 */
export type SocialMediaPostResponse = {
  /** 메인 포스트 내용 */
  post: string;

  /** 해시태그 */
  hashtags: string[];

  /** Call to Action */
  cta: string;

  /** 생성된 이미지 (include_image=true일 때만) */
  image?: GeneratedImage | null;

  /** 사용량 정보 */
  usage?: {
    tokens?: number;
    cost?: number;
  };
};

// ============================================================================
// 유틸리티 타입
// ============================================================================

/**
 * 이미지 URL 생성 헬퍼
 * Base64 데이터를 Data URL로 변환
 */
export function getImageDataUrl(image: GeneratedImage | null | undefined): string | null {
  if (!image) return null;

  if (image.type === 'base64' && image.data) {
    const format = image.format || 'png';
    return `data:image/${format};base64,${image.data}`;
  }

  if (image.type === 'url' && image.url) {
    return image.url;
  }

  return null;
}

/**
 * 이미지 유효성 검증
 */
export function isValidGeneratedImage(image: GeneratedImage | null | undefined): boolean {
  if (!image) return false;

  if (image.type === 'base64') {
    return !!image.data && image.data.length > 0;
  }

  if (image.type === 'url') {
    return !!image.url && image.url.length > 0;
  }

  return false;
}
